geo/src/algorithm/triangulate_delaunay.rs

```diff
@@ -0,0 +1,645 @@
+}
+
+impl<T: GeoFloat> TriangulateDelaunay<T> for Polygon<T>
+where
+    f64: From<T>,
+{
+    fn delaunay_triangulation(&self) -> Result<Vec<Triangle<T>>> {
+        let super_triangle = DelaunayTriangle(create_super_triangle(self)?);
+
+        let mut triangles: Vec<DelaunayTriangle<T>> = vec![super_triangle.clone()];
+        let mut processed_points: Vec<Coord<T>> = Vec::new();
+        for pt in self.exterior().coords() {
+            // If we have already added the point, do not process again
+            // to avoid producing intersecting triangles
+            if !processed_points.contains(pt) {
+                add_coordinate(&mut triangles, pt);
+                processed_points.push(*pt);
+            }
+        }
+        let triangles = remove_super_triangle(&triangles, &super_triangle);
+        Ok(triangles.iter().map(|x| x.0).collect())
+    }
+}
+
+impl<T: GeoFloat> TriangulateDelaunay<T> for MultiPoint<T>
+where
+    f64: From<T>,
+{
+    fn delaunay_triangulation(&self) -> Result<Vec<Triangle<T>>> {
+        let poly = Polygon::new(self.into_iter().cloned().collect(), vec![]);
+        poly.delaunay_triangulation()
+    }
+}
+
+fn create_super_triangle<T: GeoFloat>(geometry: &Polygon<T>) -> Result<Triangle<T>>
+where
+    f64: From<T>,
+{
+    let expand_factor = T::from(DEFAULT_SUPER_TRIANGLE_EXPANSION)
+        .ok_or(DelaunayTriangulationError::FailedToConstructSuperTriangle)?;
+    let bounds = geometry
+        .bounding_rect()
+        .ok_or(DelaunayTriangulationError::FailedToConstructSuperTriangle)?;
+    let width = bounds.width() * expand_factor;
+    let height = bounds.height() * expand_factor;
+    let bounds_min = bounds.min();
+    let bounds_max = bounds.max();
+
+    Ok(Triangle(
+        coord! {x: bounds_min.x - width, y: bounds_min.y},
+        coord! {x: bounds_max.x + width, y: bounds_min.y - height},
+        coord! {x: bounds_max.x + width, y: bounds_max.y + height},
+    ))
+}
+
+fn find_line<T: GeoFloat>(line: &Line<T>, lines: &[Line<T>]) -> Option<usize> {
+    for (idx, sample_line) in lines.iter().enumerate() {
+        if line == sample_line {
+            return Some(idx);
+        }
+    }
+    None
+}
+
+fn add_coordinate<T: GeoFloat>(triangles: &mut Vec<DelaunayTriangle<T>>, c: &Coord<T>)
+where
+    f64: From<T>,
+{
+    let mut bad_triangles: Vec<&DelaunayTriangle<T>> = Vec::new();
+
+    // Check for the triangles where the point is present within the
+    // corresponding circumcircle.
+    for tri in triangles.iter() {
+        if tri.is_in_circumcircle(c) {
+            bad_triangles.push(tri);
+        }
+    }
+
+    let mut bad_triangle_edges: Vec<Line<T>> = Vec::new();
+    let mut bad_triangle_edge_count: Vec<u128> = Vec::new();
+    for x in &bad_triangles {
+        for y in x.0.to_lines() {
+            let flipped_y = Line::new(y.end, y.start);
+            let idx =
+                find_line(&y, &bad_triangle_edges).or(find_line(&flipped_y, &bad_triangle_edges));
```

Comment:
Just a premature perf tip so feel free to ignore this maybe:

In another PR of mine, I did more or less the same (maybe that's an idication we should add this somewhere generally, but that's for another day). While reviewing, someone asked some perf questions and I tried to compare the coodinates directly, which gave me a good performance imporvement. You can read a little more about the numbers here in the commit message https://github.com/georust/geo/pull/1087/commits/49c2f040f51adbe3b37ef2cc9b367866ecc05961

I would suggest to use this here instead

```rust
bad_triangle_edges.iter().position(|other| is_line_shared(y, other))
```

and test out the performance just for fun. Btw this also avoids iterating through the vec twice.

Reply:
I used the suggestions provided as per `bad_triangle_edges.iter().find(|other| is_line_shared(y, other))` and below.  It is definitely much cleaner now!